skyportal/handlers/api/spatial_catalog.py

```diff
@@ -0,0 +1,379 @@
+    SpatialCatalog,
+    SpatialCatalogEntry,
+    SpatialCatalogEntryTile,
+)
+from ...utils.gcn import (
+    from_cone,
+    from_ellipse,
+)
+
+log = make_log('api/spatial_catalog')
+
+Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
+
+MAX_SPATIAL_CATALOG_ENTRIES = 1000
+
+
+def add_catalog(catalog_id, catalog_data):
+
+    log(f"Generating catalog with ID {catalog_id}")
+    start = time.time()
+
+    session = Session()
+    try:
+        entries = []
+        # check for cone key
+        if {'radius'}.issubset(set(catalog_data.keys())):
+            for ra, dec, name, radius in zip(
+                catalog_data['ra'],
+                catalog_data['dec'],
+                catalog_data['name'],
+                catalog_data['radius'],
+            ):
+                name = name.strip().replace(" ", "-")
+                skymap = from_cone(ra, dec, radius, n_sigma=2)
+                del skymap['localization_name']
+                skymap['entry_name'] = name
+
+                data = {'ra': ra, 'dec': dec, 'radius': radius}
+
+                entry = SpatialCatalogEntry(
+                    **{**skymap, 'catalog_id': catalog_id, 'data': data}
+                )
+                entries.append(entry)
+        elif {'amaj', 'amin', 'phi'}.issubset(set(catalog_data.keys())):
+            for ra, dec, name, amaj, amin, phi in zip(
+                catalog_data['ra'],
+                catalog_data['dec'],
+                catalog_data['name'],
+                catalog_data['amaj'],
+                catalog_data['amin'],
+                catalog_data['phi'],
+            ):
+
+                name = name.strip().replace(" ", "-")
+                skymap = from_ellipse(name, ra, dec, amaj, amin, phi)
+                skymap['entry_name'] = skymap['localization_name']
+                del skymap['localization_name']
+
+                data = {'ra': ra, 'dec': dec, 'amaj': amaj, 'amin': amin, 'phi': phi}
+
+                entry = SpatialCatalogEntry(
+                    **{**skymap, 'catalog_id': catalog_id, 'data': data}
+                )
+                entries.append(entry)
+
+        else:
+            return ValueError('Could not disambiguate keys')
+
+        session.add_all(entries)
+        session.commit()
+
+        for entry in entries:
+            tiles = [
+                SpatialCatalogEntryTile(
+                    entry_name=entry.entry_name, healpix=uniq, probdensity=probdensity
+                )
+                for uniq, probdensity in zip(entry.uniq, entry.probdensity)
+            ]
+
+            session.add_all(tiles)
+        session.commit()
+
+        flow = Flow()
+        flow.push(
+            '*',
+            "skyportal/REFRESH_SPATIAL_CATALOGS",
+        )
+
+        end = time.time()
+        duration = end - start
+
+        log(f"Generated catalog with ID {catalog_id} in {duration} seconds")
+    except Exception as e:
+        log(f"Unable to generate catalog: {e}")
```

Comment:
Maybe rollback the session here? Imagine you added the `SpatialCatalogEntry`, but then you get an error when adding the corresponding `Tile`, it would be problematic to end up with entries that are missing their tiles. In that case, it would be better to cancel the first operation.

Reply:
yes good.